internal/datastore/common/changes.go

```diff
@@ -0,0 +1,94 @@
+package common
+
+import (
+	"context"
+	"sort"
+
+	v0 "github.com/authzed/authzed-go/proto/authzed/api/v0"
+	"github.com/rs/zerolog/log"
+	"github.com/shopspring/decimal"
+
+	"github.com/authzed/spicedb/internal/datastore"
+	"github.com/authzed/spicedb/pkg/tuple"
+)
+
+// Changes represents a set of tuple mutations that are kept self-consistent
+// across one or more transaction revisions.
+type Changes map[uint64]*changeRecord
+
+type changeRecord struct {
+	tupleTouches map[string]*v0.RelationTuple
+	tupleDeletes map[string]*v0.RelationTuple
+}
+
+// AddChange adds a specific change to the complete list of tracked changes
+func (ch Changes) AddChange(
+	ctx context.Context,
+	revision uint64,
+	tpl *v0.RelationTuple,
+	op v0.RelationTupleUpdate_Operation,
+) {
+	revisionChanges, ok := ch[revision]
+	if !ok {
+		revisionChanges = &changeRecord{
+			tupleTouches: make(map[string]*v0.RelationTuple),
+			tupleDeletes: make(map[string]*v0.RelationTuple),
+		}
+		ch[revision] = revisionChanges
+	}
+
+	tplKey := tuple.String(tpl)
+
+	switch op {
+	case v0.RelationTupleUpdate_TOUCH:
+		// If there was a delete for the same tuple at the same revision, drop it
+		delete(revisionChanges.tupleDeletes, tplKey)
```

Comment:
Why the difference between if I have a delete and then touch, we're left with just the touch, but if I have touch and then delete, both will be emitted?

Reply:
The won't, the test cases confirm this.